tests/unity/CMakeLists.txt

```diff
@@ -0,0 +1,176 @@
+  get_filename_component(test_file_name "${test_file_path}" NAME)
+  set(output_file "${UNITY_PRODUCTS_DIR}/runner/runner_${test_file_name}")
+  file(MAKE_DIRECTORY "${UNITY_PRODUCTS_DIR}/runner")
+
+  execute_process(
+    COMMAND ${RUBY_EXECUTABLE}
+    ${CMOCK_DIR}/vendor/unity/auto/generate_test_runner.rb
+    ${test_file_path} ${output_file}
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+    RESULT_VARIABLE op_result
+  )
+
+  if (NOT ${op_result} EQUAL 0)
+      message(FATAL_ERROR "Failed to generate runner ${output_file}")
+  endif()
+
+  file(GLOB app_sources ${UNITY_PRODUCTS_DIR}/runner/*.c)
+  target_sources(app PRIVATE ${app_sources})
+
+  message(STATUS "Generating test runner ${unity_runner_arg3}")
+endfunction()
+
+# Generate cmock for provided header file.
+function(cmock_generate header_path dst_path)
+  get_property(CMOCK_DIR GLOBAL PROPERTY CMOCK_DIR)
+
+  file(MAKE_DIRECTORY "${dst_path}")
+
+  execute_process(
+    COMMAND ${RUBY_EXECUTABLE}
+    ${CMOCK_DIR}/lib/cmock.rb
+    --mock_prefix=mock_
+    --plugins=return_thru_ptr\;ignore_arg\;ignore\;callback\;array
+    --mock_path=${dst_path}
+    ${header_path}
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+    RESULT_VARIABLE op_result
+    OUTPUT_VARIABLE output_result
+  )
+
+  if (NOT ${op_result} EQUAL 0)
+    message(SEND_ERROR "${output_result}")
+    message(FATAL_ERROR "Failed to generate runner ${output_file}")
+  endif()
+endfunction()
+
+# Add --wrap linker option for each function listed in the input file.
+function(cmock_linker_trick func_name_path)
+  file(STRINGS ${func_name_path} contents)
+  set(linker_str "-Wl")
+  foreach(src ${contents})
+    set(linker_str "${linker_str},--wrap=${src}")
+  endforeach()
+  zephyr_link_libraries(${linker_str})
+endfunction()
+
+
+# Handle wrapping functions from mocked file.
+# Function takes header file and generates a file containing list of functions.
+# File is then passed to 'cmock_linker_trick' which adds linker option for each
+# function listed in the file.
+function(cmock_linker_wrap_trick header_file_path)
+  set(flist_file "${header_file_path}.flist")
+
+  execute_process(
+    COMMAND
+    ${PYTHON_EXECUTABLE}
+    ${NRF_DIR}/scripts/unity/func_name_list.py
+    --input ${header_file_path}
+    --output ${flist_file}
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+    RESULT_VARIABLE op_result
+    OUTPUT_VARIABLE output_result
+  )
+
+  if (NOT ${op_result} EQUAL 0)
+    message(SEND_ERROR "${output_result}")
+    message(FATAL_ERROR "Failed to parse header ${header_file_path}")
+  endif()
+  cmock_linker_trick(${flist_file})
+endfunction()
+
+# Function takes original header and prepares two version
+# - version with system calls removed and static inline functions
+#   converted to standard function declarations
+# - version with addtional __wrap_ prefix for all functions that
+#   is used to generate cmock
+function(cmock_headers_prepare in_header out_header wrap_header)
+  execute_process(
```

Comment:
Are you sure you want to use `execute_process` here ?
This means that the header file provided is ONLY processed at configure time.

When looking here: https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/93ac00157f2027b8d6516be824841e1f0bbb45b9/tests/unity/example_test/CMakeLists.txt#L16

you are providing the header to use.
Now, if a user modifies `src/foo/foo.h` and the rebuilds in this example, then `header_prepare.py` WILL NOT be executed again, and the  mocks are still based on the old header file.
`header_prepare.py` will not be executed until CMake is executed again (either manually or because of other dependencies).

If you want `header_prepare.py` to execute everytime `foo.h` is modified, then you should probably look at: 
`add_custom_command`, https://cmake.org/cmake/help/latest/command/add_custom_command.html

`add_custom_command` will also allow you to set up more dependencies for the output from `header_prepare.py` to ensure those depending on that output are updated as well.

Reply:
I've added 'add_custom_command` to generation of test runner. This is in fact very useful because adding a new test case will not require any action. Runner will be regenerated when test file is modified.

Regarding using the same approach for generation of cmocks. For now i didn't add it. It seems to me more complex as it's not only replacing `execute_process` with `add_custom_command` because cmock generation process consists of few steps, some are calling cmake functions and some are calling external command (python). It is also very unlikely that this will be needed because if we are working on writing a test for given module it's rather unlikely that we will be changing mocked API at the same time. As I would like to get that in, i considered it as not must have for now. This can be added in future if users see the need. I guess they won't be that affected.

Reply:
> It is also very unlikely that this will be needed because if we are working on writing a test for given module it's rather unlikely that we will be changing mocked API at the same time. 

Such assumptions of how user will or will not behave is like placing a trap for those working in a different way, and just waiting for them to fall into it.
Those user won't have the knowledge about this, that we have.

What if I'm working on the module that is being mocked else-where, and I as part of my changes run `build/run all unittests` (not possible in current design AFAIK), then the unit tests being built might suddenly cause build failure, due to mismatch between old mocks and updated header.
How is a user expected to be aware that `cmake` must be executed in order to get the code to compile ?

Reply:
@tejlmand i'm not saying that it's not possible. What i wanted to say is that currently i don't know how to do that. If you can help me with that i'm happy if it's added. If not then taking into account priority of this feature i would prefer to get PR merged and improve incrementally.

Reply:
@tejlmand ping. Do you agree with doing that as separate PR?

Reply:
after taking a closer look, I can see that the parts with execute process can easily be adjusted, however it also became apparent that output from one other the functions returns the functions to wrap and that information is used as argument to the linker command.
As the linker command is constructed at configure time, and does not appear to support providing a file containinig a list of functions to `ld`, then i'm fine with postponing this improvement till a later time.

Could you create a followup issue, so we don't forget ?
Priority: Low

Reply:
Issue `NCSDK-1679` created